azurelinuxagent/ga/update.py

```diff
@@ -1288,6 +1294,27 @@ def _send_heartbeat_telemetry(self, protocol):
             self._heartbeat_update_goal_state_error_count = 0
             self._last_telemetry_heartbeat = datetime.utcnow()
 
+    def _check_agent_memory_usage(self):
+        """
+        This checks the agent current memory usage and safely exit the process if agent reaches the memory limit
+        """
+        try:
+            if conf.get_enable_agent_memory_usage_check() and self._extensions_summary.converged:
+                if self._last_check_memory_usage == datetime.min or datetime.utcnow() >= (self._last_check_memory_usage + UpdateHandler.CHECK_MEMORY_USAGE_PERIOD):
```

Comment:
This would check if goal state completed and also, cgroup monitor period. 

Note: I'm using same cgroup monitor period to run this check assuming that we evaluate and calculate memory limit based on the values we get from monitoring thread.